src/builder/QuarkBuilder.sol

```diff
@@ -250,38 +262,53 @@ contract QuarkBuilder {
-    // 3. Make sure the transferIntent chain always has enough fund to cover max cost
-    function assertPaymentAffordable(
-        TransferIntent memory transferIntent,
-        Accounts.ChainAccounts[] memory chainAccountsList,
-        PaymentInfo.Payment memory payment
-    ) internal pure {
-        if (payment.isToken) {
-            uint256 totalAvailableAsset = 0;
-            for (uint256 i = 0; i < payment.maxCosts.length; ++i) {
-                uint256 paymentAssetBalanceOnChain = Accounts.sumBalances(
-                    Accounts.findAssetPositions(payment.currency, payment.maxCosts[i].chainId, chainAccountsList)
-                );
-                uint256 paymentAssetNeeded = payment.maxCosts[i].amount;
-                if (paymentAssetBalanceOnChain > paymentAssetNeeded) {
-                    totalAvailableAsset += paymentAssetBalanceOnChain - paymentAssetNeeded;
-                }
+    // Check user account can cover the cost of each actions
+    function assertActionsAffordable(Actions.Action[] memory actions, Accounts.ChainAccounts[] memory chainAccountsList)
+        internal
+        pure
+    {
+        uint256 plannedBridgeAmount = 0;
+        for (uint256 i = 0; i < actions.length; ++i) {
+            address tokenUsed;
+            uint256 amountUsed;
+            if (Strings.stringEqIgnoreCase(actions[i].actionType, Actions.ACTION_TYPE_TRANSFER)) {
+                Actions.TransferActionContext memory transferActionContext =
+                    abi.decode(actions[i].actionContext, (Actions.TransferActionContext));
+                tokenUsed = transferActionContext.token;
+                amountUsed = transferActionContext.amount;
+            } else if (Strings.stringEqIgnoreCase(actions[i].actionType, Actions.ACTION_TYPE_BRIDGE)) {
+                Actions.BridgeActionContext memory bridgeActionContext =
+                    abi.decode(actions[i].actionContext, (Actions.BridgeActionContext));
+                tokenUsed = bridgeActionContext.token;
+                amountUsed = bridgeActionContext.amount;
+                plannedBridgeAmount += amountUsed;
+            } else {
+                revert InvalidActionType();
+            }
 
-                if (payment.maxCosts[i].chainId == transferIntent.chainId) {
-                    if (paymentAssetBalanceOnChain < paymentAssetNeeded) {
+            if (tokenUsed == actions[i].paymentToken) {
+                // If the payment token is the transfer token and this is the
+                // target chain, we need to account for the transfer amount
+                uint256 paymentAssetBalanceOnChain =
+                    Accounts.sumBalances(Accounts.findAssetPositions(tokenUsed, actions[i].chainId, chainAccountsList));
+
+                // If its transfer step, check if user has enough balance to cover the transfer amount after bridge
+                if (Strings.stringEqIgnoreCase(actions[i].actionType, Actions.ACTION_TYPE_TRANSFER)) {
```

Comment:
A cleaner way to write this instead of having multiple `if (actionType == Actions.ACTION_TYPE_X)` throughout the function is to first run a filter function to generate separate lists of actions for each type.

e.g. `Actions.Action[] memory bridgeActions = findBridgeActions(actions);`

Then, you can loop through each action type and process them separately.